# 탐욕 알고리즘 실습

## 1. 집합에 대한 이해

In [ ]:
fruits = set(["avocado","tomato","banana"]) # 집합 생성
vegetables = set(["beets","carrots","tomato"]) # 집합 생성

In [ ]:
fruits | vegetables          # 합집합

{'avocado', 'banana', 'beets', 'carrots', 'tomato'}

In [ ]:
fruits & vegetables         # 교집합

{'tomato'}

In [ ]:
fruits - vegetables         # 차집합

{'avocado', 'banana'}

In [ ]:
vegetables - fruits

{'beets', 'carrots'}

## 2. 집합 커버링 문제(Set Covering Problem)

- 다음의 문제를 탐욕 알고리즘으로 풀어보시오.
    - 전국의 모든 사람들이 최소한 한 번은 쇼를 들을 수 있도록 하려면 어떤 방송국을 방문해야 할지 계산해야 합니다. 또 방송국을 방문하여 한 번 쇼를 하는데 돈이 들기 때문에 최대한 적은 수의 방송국을 돌아야 합니다. 

|라디오 방송국|청취 가능한 주|
|:---:|:---:|
|kone|nv,ut,id|
|ktwo|mt,id,wa|
|kthree|nv,ca,or|
|kfour|nv,ut|
|kfive|ca,az|

- 어느 방송국들을 가면 될까요?
- 근사 알고리즘(탐욕 알고리즘)으로 풀어봅시다.

- 우선 방송하고자 하는 주의 목록을 만듭니다

In [ ]:
states_needed = set(["mt","wa","or","id","nv","ut","ca","az"]) #집합을 이용해서 생성

In [ ]:
states_needed

{'az', 'ca', 'id', 'mt', 'nv', 'or', 'ut', 'wa'}

- 각 방송국 정보를 저장합니다.(해시 테이블사용)

In [ ]:
stations = {}
stations["kone"] = set(["id", "nv", "ut"])
stations["ktwo"] = set(["wa", "id", "mt"])
stations["kthree"] = set(["or", "nv", "ca"])
stations["kfour"] = set(["nv", "ut"])
stations["kfive"] = set(["ca", "az"])

In [ ]:
stations

{'kfive': {'az', 'ca'},
 'kfour': {'nv', 'ut'},
 'kone': {'id', 'nv', 'ut'},
 'kthree': {'ca', 'nv', 'or'},
 'ktwo': {'id', 'mt', 'wa'}}

- 해시테이블에서 `items()`함수를 살펴봅시다.
- https://www.programiz.com/python-programming/methods/dictionary/items

In [ ]:
stations.items()

dict_items([('kone', {'nv', 'ut', 'id'}), ('ktwo', {'id', 'wa', 'mt'}), ('kthree', {'or', 'nv', 'ca'}), ('kfour', {'nv', 'ut'}), ('kfive', {'az', 'ca'})])

- 여러분이 방문할 방송국의 목록을 저장할 집합이 필요합니다. 

In [ ]:
final_stations = set()

- 이제 어떤 방송국에 방송을 할지 계산하기 위한 변수를 생성 합니다.

In [ ]:
best_station = None

- 최종 코드 입니다.

In [ ]:
while states_needed: # 전체집합이 공집합이되면,
    best_station = None 
    states_covered = set()
    for station, states in stations.items(): # ('ktwo', {'mt', 'id', 'wa'} )
        covered = states_needed & states # {'ut', 'nv', 'id'}
        if len(covered) > len(states_covered):
            best_station = station   # 'kone'
            states_covered = covered # {'ut', 'nv', 'id'}
    print(states_needed)  # state_needed 의 상태를 보기위해 다음 코드를 사용합니다.
    states_needed -= states_covered  
    final_stations.add(best_station)

{'or', 'az', 'id', 'nv', 'wa', 'mt', 'ut', 'ca'}
{'or', 'az', 'wa', 'mt', 'ca'}
{'or', 'az', 'ca'}
{'az'}


In [ ]:
final_stations

{'kfive', 'kone', 'kthree', 'ktwo'}

## 3. 우리가 배운 탐욕 알고리즘을 이용해서 아래의 문제를 풀어봅시다.

- 30KG 무게의 물건 담기 (배낭 채우기)

|수확물|쌀|보리|밀|옥수수|소금|밤|조|
|---|---|---|---|---|---|---|---|
|무게|40kg|8kg|5kg|24kg|2kg|13kg|7kg|
|가격|280원|64원|60원|48원|30원|65원|56원|
|단위당가격|7|8|12|2|15|5|8|

- 물건을 쪼갤 수 없을 때 부터 생각해 봅시다. (우리가 배운 배낭 채우기)
    - W = 가방의 무게 (보통 대문자는 상수)
    - wt = 물건의 무게
    - val = 물건의 가치

In [ ]:
weight_of_items = {"쌀":40,"보리":8,"밀":5,"옥수수":24,"소금":2,"밤":13,"조":7}

In [ ]:
weight_of_items

{'밀': 5, '밤': 13, '보리': 8, '소금': 2, '쌀': 40, '옥수수': 24, '조': 7}

In [ ]:
value_of_items = [("쌀", 280), ("보리", 64),("밀", 60),("옥수수", 48),("소금", 30),("밤", 65),("조", 56)]

In [ ]:
value_of_items

[('쌀', 280),
 ('보리', 64),
 ('밀', 60),
 ('옥수수', 48),
 ('소금', 30),
 ('밤', 65),
 ('조', 56)]

In [ ]:
max(value_of_items)

('조', 56)

In [ ]:
max(value_of_items, key=lambda item:item[0]) # 0번째 열을 기반으로 인덱싱

('조', 56)

In [ ]:
max(value_of_items, key=lambda item:item[1]) # 1번째 행을 기반으로 인덱싱

('쌀', 280)

In [ ]:
max(value_of_items, key=lambda item:item[1])[0]

'쌀'

In [ ]:
max(value_of_items, key=lambda item:item[1])[1]

280

In [ ]:
weight_of_items["쌀"]

40

In [ ]:
value_of_items

[('쌀', 280),
 ('보리', 64),
 ('밀', 60),
 ('옥수수', 48),
 ('소금', 30),
 ('밤', 65),
 ('조', 56)]

In [ ]:
selected = max(value_of_items, key=lambda item:item[1])  # ('쌀', 280)

In [ ]:
selected

('쌀', 280)

In [ ]:
selected[0]

'쌀'

In [ ]:
selected[1]

280

In [ ]:
i = value_of_items.index(selected)

In [ ]:
i

0

In [ ]:
del value_of_items[i]

In [ ]:
value_of_items

[('보리', 64), ('밀', 60), ('옥수수', 48), ('소금', 30), ('밤', 65), ('조', 56)]

- Code

In [ ]:
def greedy(W, wt,val): # 배낭의 총 무게 W, 물건의 무게정보 wt, 물건의 가치 정보인 val
    # 초기 조건 설정
    knapsack = [] 
    weight = 0
    value = 0
    
    while weight < W:         # weight가 가방무게 W보다 클때까지
        selected = max(val, key=lambda item:item[1]) # ("밤", 65)
        item = selected[0]            # "밤"
        i = val.index(selected)       # 4

        if wt[item] < W - weight:  # 현재 선택한 물건을 가방에 넣을 수 있다면
            knapsack.append(selected[0])
            weight += wt[item]
            value += selected[1]
        del val[i]
        
        if len(val) == 0:  # 더이상 val 에서 볼 것이 없으면 빠져 나옴
            break
        
    return knapsack, weight, value # ex) [보리, 밀, 옥수수], 30kg, 300원

In [ ]:
weight_of_items = {"쌀":40,"보리":8,"밀":5,"옥수수":24,"소금":2,"밤":13,"조":7}

In [ ]:
value_of_items = [("쌀", 280), ("보리", 64),("밀", 60),("옥수수", 48),("소금", 30),("밤", 65),("조", 56)]

In [ ]:
W = 30

In [ ]:
greedy(W,weight_of_items,value_of_items)

(['밤', '보리', '밀', '소금'], 28, 219)

---
## 4. 최적해를 계산하기 위해서 물건을 단위당 가격으로 쪼개서 구해봅시다. 

- `value_of_items`(물건 전체의 가치)을 `unit_value_of_items`(단위당 가치)으로 변경 후 알고리즘 진행
- `fractional_greedy()` 함수를 사용하여, `OPT` (최적해)를 계산
- 위의 예에서는 266보다 더 높은 가치는 존재 하지 않음 (UPPER BOUND) 
- 우리가 앞의 `greedy()`알골리즘으로 계산한 근사해 (219 vs 266)도 충분히 좋은 해
- 다음 시간에는 물건을 위 예처럼 쪼개지 못하는 경우 최적해를 계산하는 방법에 관하여 공부합니다.

In [ ]:
def fractional_greedy(W, wt,val):
    # 초기 조건 설정
    knapsack = [] 
    weight = 0
    value = 0
    
    while weight < W:         # weight가 가방무게 W보다 클때까지
        selected = max(val, key=lambda item:item[1]) 
        item = selected[0] 
        i = val.index(selected) 

        if wt[item] < W - weight:  # 현재 선택한 물건을 가방에 넣을 수 있다면
            knapsack.append(selected[0])
            weight += wt[item]
            value += selected[1]*wt[item]
        else:
            knapsack.append(selected[0])
            weight2 = W - weight
            value += selected[1]*weight2
            break
            
        del val[i]
        
        if len(val) == 0:  # 더이상 val 에서 볼 것이 없으면 빠져 나옴
            break
        
    return knapsack, W, value # ex) [보리, 밀, 옥수수], 30kg, 300원

In [ ]:
weight_of_items = {"쌀":40,"보리":8,"밀":5,"옥수수":24,"소금":2,"밤":13,"조":7}

In [ ]:
unit_value_of_items = [("쌀", 7), ("보리", 8),("밀", 12),("옥수수", 2),("소금", 15),("밤", 5),("조", 8)]

In [ ]:
W = 30

In [ ]:
fractional_greedy(W,weight_of_items,unit_value_of_items)

(['소금', '밀', '보리', '조', '쌀'], 30, 266)